File: Java/프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.java
import java.util.ArrayDeque;

class Solution {
    public int solution(int[] stones, int k) {
        int answer = Integer.MAX_VALUE;
        
        ArrayDeque<Integer> ad = new ArrayDeque<>();
        
        for (int i = 0; i < stones.length; i++) {
            if (!ad.isEmpty() && ad.peekFirst() <= i - k) {
                ad.pollFirst();
            }
            
            while (!ad.isEmpty() && stones[ad.peekLast()] <= stones[i]) {
                ad.pollLast();
            }
            
            ad.addLast(i);
            
            if (i >= k - 1) {
                int currentWindowMax = stones[ad.peekFirst()];
                answer = Math.min(answer, currentWindowMax);
            }
        }
        
        return answer;
    }
}

File: Swift/프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.swift
import Foundation

func solution(_ stones:[Int], _ k:Int) -> Int {
var answer = Int.max
        var deque = [Int]()
        var head = 0
        
        for i in 0..<stones.count {
            if head < deque.count && deque[head] <= i - k {
                head += 1
            }
            
            while head < deque.count && stones[deque.last!] <= stones[i] {
                deque.removeLast()
            }
            
            deque.append(i)
            
            if i >= k - 1 {
                let currentWindowMax = stones[deque[head]]
                answer = min(answer, currentWindowMax)
            }
        }
        
        return answer
}